### DISCUSSION TITLE
"Autoloader optimization": Why not to use that during development?

### DISCUSSION THREAD

**Discussion by alfredbez** [composer/composer#11329] [General]
Hi there, I had all the time the impression that we should not use `-o` for `composer` commands for development, but only when we deploy things to production.

From the [composer docs](https://getcomposer.org/doc/articles/autoloader-optimization.md#autoloader-optimization):

<img width="972" alt="image" src="https://user-images.githubusercontent.com/1001186/218986948-6925c107-af42-4c19-87ce-f57f371760c0.png">

I just stumbled upon an issue in our development team yesterday, where two guys had a huge difference in performance for some API endpoints. And by "huge difference" I mean like 200ms vs 2.000ms => factor 10. We boiled it down to this flag and could reproduce both versions (the slow one and the faster one) on both machines just by running `composer dumpautoload` or `composer dumpautoload -o`. We also tried to add classes after we executed `composer dumpautoload -o` and the were found by the autoloader.

We're using macOS and our application runs inside docker containers, this surely causes these problems.

So my question is: Is that part of the documentation still valid from your point of view? Maybe it's still worth to add a section that informs users with a similar setup like mine to use that flag anyway, also during development?

**Comment by stof** [comment]
enabling level 2 optimization forces you to re-dump the autoloader each time you change the list of available classes, which is a big burden on DX.
For level 1 optimization, it will only cause issues when removing classes (or moving them to a different place) as the autoloader would still know about their old location due to the optimized map. So the burden is smaller, but it can still lead to weird errors.

**Reply by alfredbez** [reply]
Ok, I was aware of the trouble it can cause for level 2 optimization. I was also thinking about when a class was removed when using level 1 optimization and what actually will be the problem in that case, but I could not come up with a scenario (which doesn't mean there is none).

I just wanted to say, that either
* The wording of this message can be adapted because IMHO the **performance gains are** ~not~ **worth** "the trouble" [...]
* Or the message could be moved down to the level 2 optimizations
